juju/charm.py

```diff
@@ -108,17 +108,13 @@ class CharmBase(Object):
     def __init__(self, framework, key):
         super().__init__(framework, key)
 
-        for relation_name in self.framework.meta.relations:
-            relation_name = relation_name.replace('-', '_')
-            self.on.define_event(f'{relation_name}_relation_joined', RelationJoinedEvent)
-            self.on.define_event(f'{relation_name}_relation_changed', RelationChangedEvent)
-            self.on.define_event(f'{relation_name}_relation_departed', RelationDepartedEvent)
-            self.on.define_event(f'{relation_name}_relation_broken', RelationBrokenEvent)
+        self.endpoints = {}
+        for endpoint_name in self.framework.meta.endpoints:
+            self.endpoints[endpoint_name] = CharmEndpoint(self, endpoint_name)
```

Comment:
As discussed, these claim more of the attribute space within the `Charm` class (`endpoints` and `storage`), which we want to avoid.  Additionally, though, the `Object`-subclass instances in the collections use the `Charm` instance as their parent for their handles, so their `key` potentially also conflicts with other `Object`-subclass attributes of the charm.  There is the type portion in there as well, but it's still a lot of overlap and some potential for collisions that we really need to avoid.